string[] string_array
float32[] float_array
int16[] int_array
string hash # dedup